Sofa/Component/Constraint/Lagrangian/Solver/src/sofa/component/constraint/lagrangian/solver/GenericConstraintSolver.cpp

```diff
@@ -1119,6 +1074,240 @@ void GenericConstraintProblem::unbuiltGaussSeidel(SReal timeout, GenericConstrai
+    m_p.clear();
+    m_p.resize(dimension);
+
+
+    SReal *dfree = getDfree();
+    SReal *force = getF();
+    SReal **w = getW();
+    SReal tol = tolerance;
+
+    SReal *d = _d.ptr();
+
+
+    SReal error = 0.0;
+    bool convergence = false;
+    sofa::type::vector<SReal> tempForces;
+
+    if(sor != 1.0)
+    {
+        tempForces.resize(dimension);
+    }
+
+    if(scaleTolerance && !allVerified)
+    {
+        tol *= dimension;
+    }
+
+    for(int i=0; i<dimension; )
+    {
+        if(!constraintsResolutions[i])
+        {
+            msg_error("GenericConstraintSolver") << "Bad size of constraintsResolutions in GenericConstraintProblem" ;
+            break;
+        }
+        constraintsResolutions[i]->init(i, w, force);
+        i += constraintsResolutions[i]->getNbLines();
+    }
+
+    sofa::type::vector<SReal> tabErrors(dimension);
+    bool constraintsAreVerified = true;
+
+    {
+        // peform one iteration of ProjectedGaussSeidel
+        std::copy(force, force + dimension, std::begin(m_lam));
+
+        gaussSeidel_increment(false, dfree, force, w, tol, d, dimension, constraintsAreVerified, error, tabErrors);
+
+        for(int j=0; j<dimension; j++)
+        {
+            m_deltaF[j] = -(force[j] - m_lam[j]);
+            m_p[j] = - m_deltaF[j];
+        }
+    }
+
+
+
+    int iterCount = 0;
+
+    for(int i=1; i<iterationNewton; i++)
+    {
+        iterCount ++;
+        bool constraintsAreVerified = true;
+
+        for(int j=0; j<dimension; j++)
+        {
+            m_lam[j] = force[j];
+        }
+
+
+        error=0.0;
+        gaussSeidel_increment(true, dfree, force, w, tol, d, dimension, constraintsAreVerified, error, tabErrors);
+
+
+        if(allVerified)
+        {
+            if(constraintsAreVerified)
+            {
+                convergence = true;
+                break;
+            }
+        }
+        else if(error < tol) // do not stop at the first iteration (that is used for initial guess computation)
+        {
+            convergence = true;
+            break;
+        }
+
+
+        // NNCG update with the correction p
+        for(int j=0; j<dimension; j++)
+        {
+            m_deltaF_new[j] = -(force[j] - m_lam[j]);
+        }
+
+        SReal beta = m_deltaF_new.dot(m_deltaF_new) / m_deltaF.dot(m_deltaF);
+        m_deltaF.eq(m_deltaF_new, 1);
+
+        if(beta > 1)
+        {
+            m_p.clear();
+            m_p.resize(dimension);
+        }
+        else
+        {
+            for(int j=0; j<dimension; j++)
+            {
+                force[j] += beta*m_p[j];
+                m_p[j] = beta*m_p[j] - m_deltaF[j];
+            }
+        }
+    }
+
+    result_output(force, error, iterCount, convergence);
+}
+
+void GenericConstraintProblem::gaussSeidel_increment(bool measureError, SReal *dfree, SReal *force, SReal **w, SReal tol, SReal *d, int dim, bool constraintsAreVerified, SReal& error, sofa::type::vector<SReal>& tabErrors)
```

Comment:
Can be a const function

Reply:
`constraintsAreVerified` is passed by value, and you modify its value. But you don't do anything with it. I am wondering if you did not want to make it a reference, so that it is an output of the function